ksql-execution/src/main/java/io/confluent/ksql/execution/function/udtf/KudtfFlatMapper.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.function.udtf;
+
+import com.google.errorprone.annotations.Immutable;
+import io.confluent.ksql.GenericRow;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.kafka.streams.kstream.ValueMapper;
+
+/**
+ * Implements the actual flat-mapping logic - this is called by Kafka Streams
+ */
+@Immutable
+public class KudtfFlatMapper implements ValueMapper<GenericRow, Iterable<GenericRow>> {
+
+  private final TableFunctionApplier functionHolder;
+
+  public KudtfFlatMapper(final TableFunctionApplier functionHolder) {
+    this.functionHolder = functionHolder;
+  }
+
+  @Override
+  public Iterable<GenericRow> apply(final GenericRow row) {
+    final List<Object> exploded = functionHolder.apply(row);
+    final List<GenericRow> rows = new ArrayList<>(exploded.size());
+    for (Object val : exploded) {
+      final List<Object> arrayList = new ArrayList<>(row.getColumns());
+      arrayList.add(val);
```

Comment:
This is unnecessary as constructing one ArrayList from another one will copy the elements from the first one.

Reply:
It's not unnecessary, as you're adding a new element. If the capacity of the ArrayList if maxed this will result in a new array being allocated, at twice the size, and then an array copy.

Reply:
Allocation cost of small amounts of memory in Java is usually very low - this is because they are usually allocated from a chunk of memory called the TLAB that is owned by the thread. Since only that thread accesses it own TLAB memory allocation is more or less just pointer addition (no real malloc occurs).

I knocked together a quick JMH benchmark https://gist.github.com/purplefox/2ab535cd75c3f7a616f04ecc0ab3395d which supports this. 
Results are:
```
Benchmark                      Mode  Cnt   Score   Error  Units
ArrayInitBenchmark.initArray1  avgt    2  25.184          ns/op
ArrayInitBenchmark.initArray2  avgt    2  40.974          ns/op
```

This shows that pre-allocating and adding elements using addAl() is significantly slower than using the copy constructor. (iterating through them one by one is much slower still).

Small allocations aren't always bad and can often result in fewer cpu cycles than longer code paths.

But... to put this into perspective.. even if we do it the slow way, who cares? 40ns per row, means we can 25 million of these per second. So until KSQL can process around 10 million rows per second any optimisation here is not going to be remotely noticeable. I suspect that time is quite a long way off ;)